exp/runtime/internal/controllers/warmup.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1alpha1"
+	runtimeclient "sigs.k8s.io/cluster-api/internal/runtime/client"
+)
+
+const (
+	warmupTimeout  = 60 * time.Second
+	warmupInterval = 2 * time.Second
+)
+
+var _ manager.LeaderElectionRunnable = &WarmupRunnable{}
+
+// WarmupRunnable is a controller runtime LeaderElectionRunnable. It warms up the registry on controller start.
+type WarmupRunnable struct {
+	Client        client.Client
+	APIReader     client.Reader
+	RuntimeClient runtimeclient.Client
+}
+
+// NeedLeaderElection satisfies the controller runtime LeaderElectionRunnable interface.
+func (r *WarmupRunnable) NeedLeaderElection() bool {
+	return true
+}
+
+// Start attempts to warm up the registry. It will retry for 60 seconds before returning an error. An error on Start will
+// cause the CAPI controller manager to fail.
+func (r *WarmupRunnable) Start(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	ctx, cancel := context.WithTimeout(ctx, warmupTimeout)
+	defer cancel()
+
+	err := wait.PollImmediateWithContext(ctx, warmupInterval, warmupTimeout, func(ctx context.Context) (done bool, err error) {
+		if err = warmupRegistry(ctx, r.Client, r.APIReader, r.RuntimeClient); err != nil {
+			log.Error(err, "ExtensionConfig discovery failed")
```

Comment:
```suggestion
			log.Error(err, "Warmup of extension registry failed")
```
nit: I wonder if that would be better (similar in l.70). Up to you, just pick what you prefer